client/consensus/babe/src/rpc.rs

```diff
@@ -0,0 +1,235 @@
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! rpc api for babe.
+
+use std::fmt;
+use futures::{
+	prelude::*,
+	executor::ThreadPool,
+	channel::oneshot,
+	future,
+};
+use jsonrpc_core::BoxFuture;
+use jsonrpc_derive::rpc;
+use sp_consensus_babe::{AuthorityId, Epoch, AuthorityIndex, BabePreDigest};
+use crate::{SharedEpochChanges, authorship, epoch_changes::descendent_query, Config};
+use sc_keystore::KeyStorePtr;
+use std::sync::Arc;
+use sp_core::{crypto::Pair, Bytes};
+use sp_runtime::traits::{Block as BlockT, Header as _};
+use sp_consensus::{SelectChain, Error as ConsensusError};
+use sp_consensus_babe::BabeApi;
+use sp_blockchain::{HeaderBackend, HeaderMetadata, Error as BlockChainError};
+use serde::{Deserialize, Serialize};
+use sp_api::{ProvideRuntimeApi, BlockId};
+
+/// Provides rpc methods for interacting with Babe
+#[rpc(server)]
+pub trait Babe {
+	/// query slot authorship info
+	#[rpc(name = "babe_epochAuthorship")]
+	fn epoch_authorship(&self) -> BoxFuture<Vec<SlotAuthorship>>;
+}
+
+/// RPC handler for Babe
+/// provides `babe_epochAuthorship` method for querying slot authorship data.
+///
+/// Uses a background thread to calculate epoch_authorship data.
+pub struct BabeRPC<B: BlockT, C> {
+	/// shared refernce to the client.
+	client: Arc<C>,
+	/// shared reference to EpochChanges
+	shared_epoch_changes: SharedEpochChanges<B>,
+	/// shared reference to the Keystore
+	keystore: KeyStorePtr,
+	/// config (actually holds the slot duration)
+	babe_config: Config,
+	/// threadpool for spwaning cpu bound tasks.
+	threadpool: ThreadPool,
+}
+
+impl<B: BlockT, C> BabeRPC<B, C> {
+	/// creates a new instance of the BabeRpc handler.
+	pub fn new(
+		client: Arc<C>,
+		shared_epoch_changes: SharedEpochChanges<B>,
+		keystore: KeyStorePtr,
+		babe_config: Config
+	) -> Self {
+		let threadpool = ThreadPool::builder()
+			// single thread is fine.
+			.pool_size(1)
+			.create()
+			.unwrap();
+		Self {
+			client,
+			shared_epoch_changes,
+			keystore,
+			babe_config,
+			threadpool,
+		}
+	}
+}
+
+impl<B, C> Babe for BabeRPC<B, C>
+	where
+		B: BlockT,
+		C: ProvideRuntimeApi<B> + SelectChain<B> + HeaderBackend<B>
+		+ HeaderMetadata<B, Error=BlockChainError> + 'static,
+		C::Api: BabeApi<B>,
+		<<C as ProvideRuntimeApi<B>>::Api as sp_api::ApiErrorExt>::Error: fmt::Debug
+{
+	fn epoch_authorship(&self) -> BoxFuture<Vec<SlotAuthorship>> {
```

Comment:
IMO the result here should be something like:

```
{
   "pub_key1": {
      "primary": [0, 3, 10],
      "secondary": [1, 2, 8],
   },
   "pub_key2": {
      "primary": [1, 2],
      "secondary": [3, 9, 7],
   },
}
```

The first key in the result is the address of the session key that will get to claim the slot (since we try all available keys in the keystore). I think you might need to change the `claim_slot` function to return the pubkey that was used.

Reply:
cc @rphmeier WDYT?

Reply:
What are the "primary" and "secondary" fields here?

Reply:
Sorry I should have been a bit more descriptive here. the primary and secondary arrays contain the slot numbers that can be claimed by the given key, respectively for "primary" (VRF-based slots) and "secondary" slots (deterministic index-based slots).

Reply:
Yeah, that seems like a good format.